server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/filemove/FileMoveDetectionStep.java

```diff
@@ -182,64 +195,133 @@ public void visitFile(Component file) {
+    for (int removeFileIndex = 0; removeFileIndex < removedFiles.length; removeFileIndex++) {
+      ScoreMatrix.ScoreFile removedFile = removedFiles[removeFileIndex];
+      int lowerBound = (int) Math.floor(removedFile.getLineCount() * LOWER_BOUND_RATIO);
+      int upperBound = (int) Math.ceil(removedFile.getLineCount() * UPPER_BOUND_RATIO);
+      // no need to compute score if all files are out of bound, so no need to load line hashes from DB
+      if (newFiles[0].getLineCount() <= lowerBound || newFiles[lastNewFileIndex].getLineCount() >= upperBound) {
+        continue;
+      }
+      removedFilesIndexes.put(removedFile.getFileKey(), removeFileIndex);
+    }
 
+    LineHashesWithKeyDtoResultHandler rowHandler = new LineHashesWithKeyDtoResultHandler(removedFilesIndexes, removedFiles,
+      newFiles, newFileSourcesByKey, scoreMatrix);
     try (DbSession dbSession = dbClient.openSession(false)) {
-      int dbFileIndex = 0;
-      for (String removedFileKey : dbFileKeys) {
-        File fileInDb = getFile(dbSession, dtosByKey.get(removedFileKey));
-        if (fileInDb == null) {
+      for (List<String> partition : toUniqueAndSortedPartitions(removedFilesIndexes.keySet())) {
+        dbClient.fileSourceDao().scrollLineHashes(dbSession, partition, rowHandler);
+      }
+    }
+
+    return new ScoreMatrix(removedFiles, newFiles, scoreMatrix, rowHandler.getMaxScore());
+  }
+
+  private final class LineHashesWithKeyDtoResultHandler implements ResultHandler<LineHashesWithKeyDto> {
+    private final Map<String, Integer> removedFilesIndexes;
+    private final ScoreMatrix.ScoreFile[] removedFiles;
+    private final ScoreMatrix.ScoreFile[] newFiles;
+    private final Map<String, File> newFileSourcesByKey;
+    private final int[][] scoreMatrix;
+    private int maxScore;
+
+    private LineHashesWithKeyDtoResultHandler(Map<String, Integer> removedFilesIndexes, ScoreMatrix.ScoreFile[] removedFiles,
+      ScoreMatrix.ScoreFile[] newFiles, Map<String, File> newFileSourcesByKey,
+      int[][] scoreMatrix) {
+      this.removedFilesIndexes = removedFilesIndexes;
+      this.removedFiles = removedFiles;
+      this.newFiles = newFiles;
+      this.newFileSourcesByKey = newFileSourcesByKey;
+      this.scoreMatrix = scoreMatrix;
+    }
+
+    @Override
+    public void handleResult(ResultContext<? extends LineHashesWithKeyDto> resultContext) {
+      LineHashesWithKeyDto lineHashesDto = resultContext.getResultObject();
+      if (lineHashesDto.getPath() == null) {
+        return;
+      }
+      int removeFileIndex = removedFilesIndexes.get(lineHashesDto.getKey());
+      ScoreMatrix.ScoreFile removedFile = removedFiles[removeFileIndex];
+      int lowerBound = (int) Math.floor(removedFile.getLineCount() * LOWER_BOUND_RATIO);
+      int upperBound = (int) Math.ceil(removedFile.getLineCount() * UPPER_BOUND_RATIO);
+
+      for (int newFileIndex = 0; newFileIndex < newFiles.length; newFileIndex++) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Reduce the total number of break and continue statements in this loop to use at most one. ([squid:S135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Fsl%2FSONAR-10253%2Foptimize_file_move_detection_based_on_file_size&issues=AWHRfxNUNjvrBDlUoofF&open=AWHRfxNUNjvrBDlUoofF)

[ISSUE_KEY]: AWHRfxNUNjvrBDlUoofF